toolkit/featureforms/src/main/java/com/arcgismaps/toolkit/featureforms/components/text/FormTextFieldState.kt

```diff
@@ -85,84 +107,257 @@ internal class FormTextFieldState(
-        context.getString(R.string.maximum_n_chars, maxLength)
-    } else {
-        context.getString(R.string.maximum_n_chars, 254)
-        // TODO: when consuming the core API throw here and remove the line above.
-        //throw IllegalStateException("invalid form data or attribute: text field must have a nonzero max length")
-    }
 
     init {
         scope.launch {
             value.drop(1).collect { newValue ->
-                if (isEditable.value) {
-                    validate(newValue)
-                }
+                updateValidation(newValue)
             }
         }
         scope.launch {
-            isFocused.collect {
+            isRequired.drop(1).collect {
+                updateValidation(value.value)
+            }
+        }
+        scope.launch {
+            isFocused.drop(1).collect {
                 if (it) {
-                    validate(value.value)
+                    wasFocused = true
                 }
+                updateValidation(value.value)
             }
         }
     }
-
-    /**
-     * Validates the current [value]'s length based on the [minLength], [maxLength], and [isRequired] and sets the
-     * [hasError] and [_errorMessage] if there was an error in validation.
-     */
-    private fun validate(value: String) {
-        _hasError.value = if (isRequired.value && value.isEmpty()) {
-            _errorMessage = context.getString(R.string.required)
-            true
-        } else if (value.length !in minLength..maxLength) {
-            _errorMessage = helperText
-            true
+    
+    private fun updateValidation(value: String) {
+        validationErrors = validate(value)
+        val errorToDisplay = errorMessageToDisplay(value, validationErrors)
+        _supportingTextIsErrorMessage.value = errorToDisplay != NoError
+        _supportingText.value = if (errorToDisplay != NoError) {
+            errors[errorToDisplay] ?: throw IllegalStateException("validation error must have a message")
         } else {
-            false
+            description.ifEmpty {
+                if (_isFocused.value) helperText else ""
+            }
         }
+        _hasError.value = validationErrors.isNotEmpty()
     }
-
+    
+    private fun validateTextRange(value: String): ValidationErrorState =
+        if (value.length !in minLength..maxLength) {
+            if (minLength > 0 && maxLength > 0) {
+                if (minLength == maxLength) {
+                    ExactCharConstraint
+                } else {
+                    MinMaxCharConstraint
+                }
+            } else {
+                MaxCharConstraint
+            }
+        } else {
+            NoError
+        }
+    
+    private fun validateNumericRange(value: String): ValidationErrorState =
+        if (domain != null && domain is RangeDomain) {
+            if (fieldType.isIntegerType) {
+                val (min, max) = domain.asLongTuple
+                val numberVal = value.toLong()
```

Comment:
```suggestion
                val numberVal = value.toLongOrNull()
```
This could throw an exception if the input is not a number.